api/src/main/java/com/okta/idx/sdk/api/wrapper/AuthenticationWrapper.java

```diff
@@ -0,0 +1,264 @@
+
+            // Check if identify flow needs to include credentials
+            boolean isIdentifyInOneStep = isRemediationRequireCredentials(RemediationType.IDENTIFY, introspectResponse);
+
+            IdentifyRequest identifyRequest;
+
+            if (isIdentifyInOneStep) {
+                Credentials credentials = new Credentials();
+                credentials.setPasscode(authenticationOptions.getPassword().toCharArray());
+
+                identifyRequest = IdentifyRequestBuilder.builder()
+                        .withIdentifier(authenticationOptions.getUsername())
+                        .withCredentials(credentials)
+                        .withStateHandle(stateHandle)
+                        .build();
+            } else {
+                identifyRequest = IdentifyRequestBuilder.builder()
+                        .withIdentifier(authenticationOptions.getUsername())
+                        .withStateHandle(stateHandle)
+                        .build();
+            }
+
+            // identify user
+            IDXResponse identifyResponse = remediationOption.proceed(client, identifyRequest);
+
+            if (isIdentifyInOneStep) {
+                // we expect success
+                if (!identifyResponse.isLoginSuccessful()) {
+                    // verify if password expired
+                    if (isRemediationRequireCredentials(RemediationType.REENROLL_AUTHENTICATOR, identifyResponse)) {
+                        logger.warn("Password expired!");
+                        authenticationResponse.setAuthenticationStatus(AuthenticationStatus.PASSWORD_EXPIRED);
+                    } else {
+                        logger.error("Unexpected remediation {}", RemediationType.REENROLL_AUTHENTICATOR);
+                        Arrays.stream(identifyResponse.getMessages().getValue()).forEach(msg -> authenticationResponse.addError(msg.getMessage()));
+                        authenticationResponse.setAuthenticationStatus(AuthenticationStatus.FAILURE);
+                    }
+                } else {
+                    // login successful
+                    logger.info("Login Successful!");
+                    tokenResponse = identifyResponse.getSuccessWithInteractionCode().exchangeCode(client, idxClientContext);
+                    authenticationResponse.setAuthenticationStatus(AuthenticationStatus.SUCCESS);
+                    authenticationResponse.setTokenResponse(tokenResponse);
+                }
+            } else {
+                if (!isRemediationRequireCredentials(RemediationType.CHALLENGE_AUTHENTICATOR, identifyResponse)) {
+                    logger.error("Unexpected remediation {}", RemediationType.CHALLENGE_AUTHENTICATOR);
+                    Arrays.stream(identifyResponse.getMessages().getValue()).forEach(msg -> authenticationResponse.addError(msg.getMessage()));
+                    authenticationResponse.setAuthenticationStatus(AuthenticationStatus.FAILURE);
```

Comment:
Curious about this approach. Why returning failure state instead of just throwing exceptions? I'm currently throwing exceptions, but interested to learn your reasons to see if I should do the same.

Reply:
I’m following this approach because we can easily accumulate all the errors inside `Messages` in a generic way into this list. The “throw exception” approach is fine too, but we need to type it to a correct category (a single `RemediationUnknownException` type may not fit all error message types).

I mean.. `messages` might have any kind of error returned. So we need to define new exception type to correctly categorize the different possible errors. If it were a simple list of string errors, I thought it’d be easier to maintain.